mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -2390,85 +2525,152 @@ struct GridwiseGemmRewritePattern : public OpRewritePattern<miopen::GridwiseGemm
-                       "and should not haven been attempted");
+      llvm::errs() << "Vector loads/stores aren't possible in the G dimension "
+                      "and should not haven been attempted";
+      return failure();
     }
 
     // llvm::errs() << "thread slice lengths for Matrix A\n";
     // llvm::errs() << GemmABlockCopyThreadSliceLengths_GemmK << " ";
-    // llvm::errs() << GemmABlockCopyThreadSliceLengths_GemmM << "\n";
+    // llvm::errs() << GemmABlockCopyThreadSliceLengths_GemmM << " ";
+    // llvm::errs() << GemmABlockCopyThreadSliceLengths_GemmKPack << "\n";
+
+    if (GemmABlockCopyThreadSliceLengths_GemmK == 0 ||
+        GemmABlockCopyThreadSliceLengths_GemmM == 0 ||
+        GemmABlockCopyThreadSliceLengths_GemmKPack == 0) {
+      llvm::errs() << "Blockwise copy slice lengths for matrix A is zero which "
+                      "is invalid.\n";
+      return failure();
+    }
 
     // Compute ThreadClusterLengths for Matrix A.
-    int64_t GemmABlockCopyClusterLengths_GemmK =
+    uint64_t GemmABlockCopyClusterLengths_GemmKPack =
+        KPack / GemmABlockCopyThreadSliceLengths_GemmKPack;
+    uint64_t GemmABlockCopyClusterLengths_GemmK =
         KPerBlock / GemmABlockCopyThreadSliceLengths_GemmK;
     // int64_t GemmABlockCopyClusterLengths_GemmM =
     //    MPerBlock / GemmABlockCopyThreadSliceLengths_GemmM;
 
     // llvm::errs() << "thread cluster lengths for Matrix A\n";
     // llvm::errs() << GemmABlockCopyClusterLengths_GemmK << " ";
-    // llvm::errs() << GemmABlockCopyClusterLengths_GemmM << "\n";
+    // llvm::errs() << GemmABlockCopyClusterLengths_GemmM << " ";
+    // llvm::errs() << GemmABlockCopyClusterLengths_GemmKPack << "\n";
 
     // Compute ThreadSliceLengths for Matrix B.
     int64_t GemmBBlockCopyNumberDataPerThread =
         NPerBlock * KPerBlock / BlockSize;
 
     int64_t GemmBBlockCopyThreadSliceLengths_GemmK;
     int64_t GemmBBlockCopyThreadSliceLengths_GemmN;
+    int64_t GemmBBlockCopyThreadSliceLengths_GemmKPack = 1;
     assert(matrix_b_source_vector_read_dim != GemmG);
     switch (matrix_b_source_vector_read_dim) {
     case GemmK:
-      GemmBBlockCopyThreadSliceLengths_GemmK = matrix_b_source_data_per_read;
-      GemmBBlockCopyThreadSliceLengths_GemmN =
-          GemmBBlockCopyNumberDataPerThread /
-          GemmBBlockCopyThreadSliceLengths_GemmK;
+      if (KPack > 1) {
+        GemmBBlockCopyThreadSliceLengths_GemmKPack =
+            matrix_b_source_data_per_read;
+        GemmBBlockCopyThreadSliceLengths_GemmK =
+            KPack / matrix_b_source_data_per_read;
+        GemmBBlockCopyThreadSliceLengths_GemmN =
+            GemmBBlockCopyNumberDataPerThread / KPack;
+      } else {
+        GemmBBlockCopyThreadSliceLengths_GemmK = matrix_b_source_data_per_read;
+        GemmBBlockCopyThreadSliceLengths_GemmN =
+            GemmBBlockCopyNumberDataPerThread /
+            GemmBBlockCopyThreadSliceLengths_GemmK;
+      }
```

Comment:
As a general question ... what would happen if, instead of having all this `if (kPack > 1)` stuff, you just always added a kPack dimension that was frequently length 1? Would we expect that to lead to worse ISA?

Reply:
We don't know that yet. It could be addressed in a future PR to simplify the logic.

Reply:
Fair enough, future PR it is - this wasn't a critical complaint, just an observation. Do we want a ticket for the investigation?